### PR TITLE
[hw,csrng,sv] Update coverpoints for CSRNG

### PR DESCRIPTION
**Pull request by crteja**
- Covergroup updates
  - Update coverpoints for `acmd` to cover all values supported by the field
  - Update `cp_flags` to include invalid values of `flags0` field
  - Add a covergroup for fatal and recover alerts
  - Add a covergroup for efuse tieoff `otp_en_sw_app_read`
  - Add a covergroup for fips bit in `genbits_vld` register
  - Add sample calls for newly added covergroups in scoreboard

- Remove `TODO` that get resolved by newly added covergroups
- Update covergroup descriptions in testplan

This PR resolves https://github.com/lowRISC/opentitan/issues/17418 by adding covergroups for TODO comments

### CODE REVIEW

File: hw/ip/csrng/data/csrng_testplan.hjson
Comment (marnovandermaas) [inline hw/ip/csrng/data/csrng_testplan.hjson:190]:
```suggestion
            - cp_sw_app_read : Covers for values supported by otp_en_csrng_sw_app_read
```

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (marnovandermaas) [inline hw/ip/csrng/dv/cov/csrng_cov_if.sv:290]: `INV` as well as `GENB` and `GENU` are commands that are used internally in the CSRNG FSM. It is illegal to receive these from an external source.

**Review by marnovandermaas — COMMENTED** [review]
Thanks for working on this. Some Initial comments from my side.

File: hw/ip/csrng/data/csrng_testplan.hjson
Comment (crteja) [reply hw/ip/csrng/data/csrng_testplan.hjson]: Fixed typo in [8cb28ae](https://github.com/lowRISC/opentitan/compare/b75480112ac0bbf7313fda83589a19f3f96280c2..8cb28ae611d26d1f9c660443ec2f2cf790dedbc3)

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (vogelpi) [inline hw/ip/csrng/dv/cov/csrng_cov_if.sv:295]: Is there a particular reason for not using the enum values defined in `csrng_pkg.sv`:
```
  typedef enum logic [2:0] {
    INV  = 3'h0,
    INS  = 3'h1,
    RES  = 3'h2,
    GEN  = 3'h3,
    UPD  = 3'h4,
    UNI  = 3'h5,
    GENB = 3'h6,
    GENU = 3'h7
  } acmd_e;
```
In particular, this would facilitate future code changes.

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (vogelpi) [inline hw/ip/csrng/dv/cov/csrng_cov_if.sv:388]: Nit: typo Coerpoint -> Coverpoint

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (vogelpi) [inline hw/ip/csrng/dv/cov/csrng_cov_if.sv:398]: I am not 100% but I believe that besides read_int_state_val_reg` this will try to cross `efuse_sw_app_enable` which can be `MuBi8True` and `MuBi8False` with `read_int_state` which can be anything between 0 and 15. I see the following two issues with that:
1. `otp_en_cs_sw_app_read` could also have an an invalid encoding, probably need to update `cp_sw_app_read` accordingly.
2. It's going to be very hard and not meaningful to hit all 16 cases of `read_int_state`. What we should do in my view is to just look distinguish between `MuBi4True`, `MuBi4False` and invalid encodings.

What do you think?

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (vogelpi) [inline hw/ip/csrng/dv/cov/csrng_cov_if.sv:401]: Same comment here.

File: hw/ip/csrng/data/csrng_testplan.hjson
Comment (vogelpi) [inline hw/ip/csrng/data/csrng_testplan.hjson:74]: Have you checked whether this is already tested? My understanding is that this PR "just" adds the missing coverpoints. In case the features aren't being tested yet, this will show in the corresponding coverpoints not being hit. Right?

Comment (vogelpi) [reply hw/ip/csrng/data/csrng_testplan.hjson:74]: Specifically I am bit worried about the FIPS bit. Because the now added coverpoint will tell us if tests read both output data generated from FIPS compliant and in-compliant seeds. But it won't show us whether after reading FIPS compliant bits once, FIPS in-compliant bits can be read again. I.e. whether the FIPS bit going low at the input eventually becomes visible at the output again. This is pretty important from a security perspective. Have you insight into whether is tested somehow?

File: hw/ip/csrng/data/csrng_testplan.hjson
Comment (vogelpi) [inline hw/ip/csrng/data/csrng_testplan.hjson:113]: The [CTRL register](https://opentitan.org/book/hw/ip/csrng/data/csrng.html#ctrl) has three fields: READ_INT_STATE, SW_APP_ENABLE and ENABLE. The last one is the overall module enable for CSRNG and I think this is still missing and should be added to the covergroup.

**Review by vogelpi — COMMENTED** [review]
Thanks @crteja for working on this!

I have a couple of suggestions/questions, please see the detailed comments.

File: hw/ip/csrng/data/csrng_testplan.hjson
Comment (crteja) [reply hw/ip/csrng/data/csrng_testplan.hjson:74]: > Have you checked whether this is already tested? My understanding is that this PR "just" adds the missing coverpoints. In case the features aren't being tested yet, this will show in the corresponding coverpoints not being hit. Right?

Yes, this change only adds coverpoints.

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (crteja) [reply hw/ip/csrng/dv/cov/csrng_cov_if.sv:398]: 1. Yes, will update `otp_en_cs_sw_app_read` to cover invalid values
2. Got it. Will update `cp_read_int_state` and use the coverpoint in cross

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (crteja) [reply hw/ip/csrng/dv/cov/csrng_cov_if.sv]: used enum values in bins

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (crteja) [reply hw/ip/csrng/dv/cov/csrng_cov_if.sv]: Fixed type

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (crteja) [reply hw/ip/csrng/dv/cov/csrng_cov_if.sv]: Updated coverpoints as per comments

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (crteja) [reply hw/ip/csrng/dv/cov/csrng_cov_if.sv]: Updated coverpoints as per comments

File: hw/ip/csrng/data/csrng_testplan.hjson
Comment (crteja) [reply hw/ip/csrng/data/csrng_testplan.hjson:113]: Added a bin for `ctrl`.`enable` using the `enable` variable in `csrng_env_cfg.sv`

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (vogelpi) [inline hw/ip/csrng/dv/cov/csrng_cov_if.sv:179]: This should be `MuBi4...` instead of `MuBi8...`.

File: hw/ip/csrng/data/csrng_testplan.hjson
Comment (vogelpi) [reply hw/ip/csrng/data/csrng_testplan.hjson:74]: In the interest of time, let me create an issue for that. Then we can move ahead and proceed with this PR.

**Review by vogelpi — APPROVED** [review]
LGTM, some minor nits remaining.

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (vogelpi) [reply hw/ip/csrng/dv/cov/csrng_cov_if.sv]: Thanks for changing this.

Ah wait, now the cross is without `otp_en_cs_sw_app_read`. Shouldn't this be included as well? Sorry if my previous comment was unclear and you've removed it based on that. This is not what I meant.

File: hw/ip/csrng/data/csrng_testplan.hjson
Comment (marnovandermaas) [reply hw/ip/csrng/data/csrng_testplan.hjson:74]: So in terms of toggle coverage, FIPS bit going low from entropy source is covered:
image

I am not aware of any test that tests the proper behavior here though: i.e. when entropy source provides a non-FIPS value the value that CSRNG outputs is also non-FIPS.

**Review by marnovandermaas — COMMENTED** [review]
Do you have results of running a nightly on top of this PR?

**Comment by crteja** [comment]
> Do you have results of running a nightly on top of this PR?

no, I haven't run nightly with these changes. I can give a run now.

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (crteja) [reply hw/ip/csrng/dv/cov/csrng_cov_if.sv:179]: Fixed typo

File: hw/ip/csrng/dv/cov/csrng_cov_if.sv
Comment (crteja) [reply hw/ip/csrng/dv/cov/csrng_cov_if.sv]: Updated crosses with comments describing the intended scenario [9da23c](https://github.com/lowRISC/opentitan/compare/82a9dd349fee2c3eb5851e232b884c1926475891..9da23ce86bb2220179e028b4db8bac2c570b0050)

Comment (vogelpi) [reply hw/ip/csrng/dv/cov/csrng_cov_if.sv]: Cool, thanks @crteja !

**Comment by crteja** [comment]
Functional Coverage in `csrng_cov_if` went down with the change

latest nightly report:
image

After update:
image

**Comment by marnovandermaas** [comment]
I removed the `csrng_alert_cg` because this is already covered by the toggle coverage for raising a recoverable or a fatal alert and the existing `csrng_recov_alert_sts_cg` and `csrng_err_code_cg`.

**Comment by marnovandermaas** [comment]
I added more ignore bins to the crosses so that we don't have an explosion of functional coverpoints that are unlikely to be hit.

**Comment by marnovandermaas** [comment]
Rebase to get around RTL block

**Comment by marnovandermaas** [comment]
Quite a few force pushes because I was trying to get the spacing correct.

**Review by marnovandermaas — APPROVED** [review]
After my changes, this PR looks good to me. Let's wait for CI to pass before merging.

**Pull request merged**
